src/style/style_layer.js

```diff
@@ -245,8 +245,11 @@ class StyleLayer extends Evented {
     }
 
     isStateDependent() {
-        for (const property in (this: any).paint._values) {
-            const value = (this: any).paint.get(property);
+        if (!this.paint)
+            return false;
```

Comment:
Why is `paint` (or `layout`) optional for this `StyleLayer`?

Reply:
Because it's calculated on `recalculate`, not in constructor. Alternatively, we can try setting it to empty objects there, but I think the PR solution is fine.

Reply:
Hm, I think we should investigate this further rather than covering up the symptoms. Returning `false` from `isStateDependendent` sounds wrong (since it's in an undefined state). Instead, we should not call `isStateDependent` when `recalculate` has never been called, or ensure that it is called in the constructor.

Reply:
> we should not call isStateDependent when recalculate has never been called, or ensure that it is called in the constructor.

I wasn't able to reproduce this error condition even after a bunch of experimentation. In [36270d](https://github.com/mapbox/mapbox-gl-js/commit/36270d8381b2c30d40ef75a26e330beaf87ea4af) I took the approach of moving the paint property evaluation to the worker. This should have the additional benefit of not needing to parse the expressions on the main thread.